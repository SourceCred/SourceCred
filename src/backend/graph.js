// @flow

export type Address = {
  repositoryName: string,
  pluginName: string,
  id: string,
};

export type Node<T> = {
  address: Address,
  inEdges: Address[],
  outEdges: Address[],
  payload: T,
};

export type Edge<T> = {
  address: Address,
  src: Address,
  dst: Address,
  payload: T,
};

export type Graph = {
  nodes: {[stringAddress: string]: Node<mixed>},
  edges: {[stringAddress: string]: Edge<mixed>},
};

export function addressToString(address: Address) {
  if (address.repositoryName.includes("$")) {
    const escaped = JSON.stringify(address.repositoryName);
    throw new Error(`address.repositoryName must not include "\$": ${escaped}`);
  }
  if (address.pluginName.includes("$")) {
    const escaped = JSON.stringify(address.pluginName);
    throw new Error(`address.pluginName must not include "\$": ${escaped}`);
  }
  if (address.id.includes("$")) {
    const escaped = JSON.stringify(address.id);
    throw new Error(`address.id must not include "\$": ${escaped}`);
  }
  return `${address.repositoryName}\$${address.pluginName}\$${address.id}`;
}

export function stringToAddress(string: string) {
  const parts = string.split("$");
  if (parts.length !== 3) {
    const escaped = JSON.stringify(string);
    throw new Error(`Input should have exactly two \$s: ${escaped}`);
  }
  return {
    repositoryName: parts[0],
    pluginName: parts[1],
    id: parts[2],
  };
}
