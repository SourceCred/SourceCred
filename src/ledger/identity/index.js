// @flow

export type {IdentityId} from "./id";
export type {Identity} from "./identity";
export {
  newIdentity,
  contractions,
  graphNode,
  parser as identityParser,
} from "./identity";

export type {IdentityType} from "./identityType";

export type {Name} from "./name";
export {nameFromString, parser as nameParser} from "./name";

export type {Alias} from "./alias";
export {parser as aliasParser} from "./alias";

export {declaration} from "./declaration";
