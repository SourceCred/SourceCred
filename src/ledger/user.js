// @flow

/**
 * This module has a core data type identifying SourceCred users.
 *
 * The scope for this data type is to model:
 * - a unique identifier for each user
 * - a unique (renameable) username they choose
 * - the address of every node they correspond to in the graph
 *
 * Unlike most other state in SourceCred, the User state is
 * nondeterministically generated by SourceCred itself, and then persisted
 * long-term within the instance.
 *
 * This is in contrast to Graph data, which usually comes from an external
 * source, and is not persisted long-term, but instead is re-generated when
 * needed.
 *
 * In particular, this kernel of user data is stored within the core ledger,
 * since it's necessary to track consistently when tracking Grain distribution.
 * This type should not grow to include all the data that the UI will
 * eventually want to show; that should be kept in a different data store which
 * isn't being used as a transaction ledger.
 *
 */
import {type Uuid, parser as uuidParser} from "../util/uuid";
import * as C from "../util/combo";
import {
  type NodeAddressT,
  NodeAddress,
  type Node as GraphNode,
} from "../core/graph";

/**
 * We validate usernames using GitHub-esque rules.
 *
 * Usernames are always lower-case.
 */
export opaque type Username: string = string;
const USERNAME_PATTERN = /^[A-Za-z0-9-]+$/;

export type UserId = Uuid;
export type User = {|
  // UUID, assigned when the user is created.
  +id: UserId,
  +name: Username,
  // Every other node in the graph that this user corresponds to.
  // Does not include the user's "own" address, i.e. the result
  // of calling (userAddress(user.id)).
  +aliases: $ReadOnlyArray<NodeAddressT>,
|};

// It's not in the typical [owner, name] format because it isn't provided by a plugin.
// Instead, it's a raw type owned by SourceCred project.
export const USER_PREFIX = NodeAddress.fromParts([
  "sourcecred",
  "core",
  "USER",
]);

/**
 * Parse a Username from a string.
 *
 * Throws an error if the username is invalid.
 */
export function usernameFromString(username: string): Username {
  if (!username.match(USERNAME_PATTERN)) {
    throw new Error(`invalid username: ${username}`);
  }
  return username.toLowerCase();
}

export function userAddress(id: UserId): NodeAddressT {
  return NodeAddress.append(USER_PREFIX, id);
}

export function graphNode({id, name}: User): GraphNode {
  return {
    address: userAddress(id),
    description: name,
    timestampMs: null,
  };
}

export const usernameParser: C.Parser<Username> = C.fmap(
  C.string,
  usernameFromString
);

export const userParser: C.Parser<User> = C.object({
  id: uuidParser,
  name: usernameParser,
  aliases: C.array(NodeAddress.parser),
});
