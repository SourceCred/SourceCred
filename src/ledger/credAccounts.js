// @flow

/**
 * This module outputs aggregated data that combines Cred Scores with Ledger
 * Account data.
 *
 * We use this internally when creating Grain distributions using a Ledger and
 * a Cred View. It's also an experimental output format which gives overall
 * information on the cred in an instance. We may remove it or make breaking
 * changes to it in the future.
 */
import {sum} from "d3-array";
import {Ledger, type Account} from "./ledger";
import {CredView} from "../analysis/credView";
import {type TimestampMs} from "../util/timestamp";
import {NodeAddress, type NodeAddressT} from "../core/graph";

export type Cred = $ReadOnlyArray<number>;

export type CredAccount = {|
  +cred: Cred,
  +totalCred: number,
  +account: Account,
|};

export type UnclaimedAlias = {|
  +address: NodeAddressT,
  // We include the description for convenience in figuring out who this user is,
  // rendering in a UI, etc. This is just the description from the Graph.
  +description: string,
  +totalCred: number,
  +cred: Cred,
|};

export type CredAccountData = {|
  // Regular accounts: an identity with Cred, and potentially Grain
  +accounts: $ReadOnlyArray<CredAccount>,
  // Unclaimed aliases: An account on some platform that hasn't yet been
  // connected to any SourceCred identity
  +unclaimedAliases: $ReadOnlyArray<UnclaimedAlias>,
  // The timestamps demarcating the ends of the Cred intervals.
  // For interpreting the Cred data associated with cred accounts and
  // unclaimed accounts.
  +intervalEndpoints: $ReadOnlyArray<TimestampMs>,
|};

export function computeCredAccounts(
  ledger: Ledger,
  credView: CredView
): CredAccountData {
  const grainAccounts = ledger.accounts();
  const userlikeInfo = new Map();
  for (const {address, credOverTime, description} of credView.userNodes()) {
    if (credOverTime == null) {
      throw new Error(
        `userlike ${NodeAddress.toString(address)} does not have detailed cred`
      );
    }
    userlikeInfo.set(address, {cred: credOverTime.cred, description});
  }
  const intervalEndpoints = credView.credResult().credData.intervalEnds;
  return _computeCredAccounts(grainAccounts, userlikeInfo, intervalEndpoints);
}

export function _computeCredAccounts(
  grainAccounts: $ReadOnlyArray<Account>,
  userlikeInfo: Map<NodeAddressT, {|+cred: Cred, +description: string|}>,
  intervalEndpoints: $ReadOnlyArray<TimestampMs>
): CredAccountData {
  const aliases: Set<NodeAddressT> = new Set();
  const accountAddresses: Set<NodeAddressT> = new Set();

  const accounts = [];
  const unclaimedAliases = [];

  for (const account of grainAccounts) {
    accountAddresses.add(account.identity.address);
    for (const alias of account.identity.aliases) {
      aliases.add(alias);
    }
    const info = userlikeInfo.get(account.identity.address);
    if (info == null) {
      throw new Error(
        `cred sync error: no info for account ${account.identity.name}`
      );
    }
    const {cred} = info;
    const credAccount = {account, cred, totalCred: sum(cred)};
    accounts.push(credAccount);
  }

  for (const [userAddress, info] of userlikeInfo.entries()) {
    if (accountAddresses.has(userAddress)) {
      // This userlike actually has an explicit account
      continue;
    }
    if (aliases.has(userAddress)) {
      throw new Error(
        `cred sync error: alias ${NodeAddress.toString(
          userAddress
        )} included in Cred scores`
      );
    }
    const {cred, description} = info;
    unclaimedAliases.push({
      address: userAddress,
      cred,
      totalCred: sum(cred),
      description,
    });
  }
  return {accounts, unclaimedAliases, intervalEndpoints};
}
