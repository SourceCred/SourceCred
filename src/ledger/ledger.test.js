// @flow

import deepFreeze from "deep-freeze";
import cloneDeep from "lodash.clonedeep";
import {random as randomUuid} from "../util/uuid";
import {NodeAddress} from "../core/graph";
import {Ledger, type DistributionPolicy} from "./ledger";
import {userAddress} from "./user";
import * as G from "./grain";

describe("ledger/ledger", () => {
  // Helper for constructing Grain values.
  const g = (s) => G.fromString(s);
  function setFakeDate(ts: number) {
    jest.spyOn(global.Date, "now").mockImplementationOnce(() => ts);
  }

  // Verify that a method fails, throwing an error, without mutating the ledger.
  function failsWithoutMutation(
    ledger: Ledger,
    operation: (Ledger) => any,
    message: string
  ) {
    const copy = cloneDeep(ledger);
    expect(() => operation(ledger)).toThrow(message);
    expect(copy).toEqual(ledger);
  }

  const a1 = NodeAddress.fromParts(["a1"]);
  const a2 = NodeAddress.fromParts(["a2"]);

  describe("user updates", () => {
    describe("createUser", () => {
      it("works", () => {
        setFakeDate(123);
        const l = new Ledger();
        const id = l.createUser("foo");
        const foo = l.userById(id);
        expect(l.users()).toEqual([foo]);
        expect(l.actionLog()).toEqual([
          {
            type: "CREATE_USER",
            username: "foo",
            version: 1,
            timestamp: 123,
            userId: id,
          },
        ]);
      });
      it("throws an error if the username is invalid", () => {
        const ledger = new Ledger();
        const thunk = () => ledger.createUser("foo bar");
        failsWithoutMutation(ledger, thunk, "invalid username");
        expect(ledger.users()).toEqual([]);
      });
      it("throws an error if the username is taken", () => {
        const ledger = new Ledger();
        ledger.createUser("foo");
        const thunk = () => ledger.createUser("foo");
        failsWithoutMutation(ledger, thunk, "username already taken");
      });
      it("throws on unrecognized version", () => {
        // $FlowExpectedError
        expect(() => new Ledger()._createUser({version: 1337})).toThrowError(
          "unrecognized version"
        );
      });
    });

    describe("renameUser", () => {
      it("works", () => {
        const ledger = new Ledger();
        setFakeDate(0);
        const id = ledger.createUser("foo");
        setFakeDate(1);
        ledger.renameUser(id, "bar");
        const user = ledger.userById(id);

        expect(user).toEqual({id, name: "bar", aliases: []});
        expect(ledger.userByUsername("bar")).toEqual(user);
        expect(ledger.userByUsername("foo")).toEqual(undefined);
        expect(ledger.users()).toEqual([user]);

        expect(ledger.actionLog()).toEqual([
          {
            type: "CREATE_USER",
            version: 1,
            username: "foo",
            timestamp: 0,
            userId: id,
          },
          {
            type: "RENAME_USER",
            version: 1,
            newName: "bar",
            userId: id,
            timestamp: 1,
          },
        ]);
      });
      it("fails if the user already has that name", () => {
        const ledger = new Ledger();
        const id = ledger.createUser("foo");
        const thunk = () => ledger.renameUser(id, "foo");
        failsWithoutMutation(
          ledger,
          thunk,
          "renameUser: user already has name"
        );
      });
      it("fails on nonexistent user id", () => {
        const ledger = new Ledger();
        failsWithoutMutation(
          ledger,
          (l) => l.renameUser(randomUuid(), "bar"),
          "renameUser: no user matches id"
        );
      });
      it("fails on username conflict", () => {
        const ledger = new Ledger();
        const fooId = ledger.createUser("foo");
        ledger.createUser("bar");
        const thunk = () => ledger.renameUser(fooId, "bar");
        failsWithoutMutation(
          ledger,
          thunk,
          "renameUser: conflict on username bar"
        );
      });
      it("fails on invalid username", () => {
        const ledger = new Ledger();
        const fooId = ledger.createUser("foo");
        const thunk = () => ledger.renameUser(fooId, "foo bar");
        failsWithoutMutation(ledger, thunk, "invalid username");
      });
      it("throws on unrecognized version", () => {
        // $FlowExpectedError
        expect(() => new Ledger()._renameUser({version: 1337})).toThrowError(
          "unrecognized version"
        );
      });
    });

    describe("addAlias", () => {
      it("works", () => {
        const ledger = new Ledger();
        setFakeDate(0);
        const id = ledger.createUser("foo");
        setFakeDate(1);
        ledger.addAlias(id, a1);
        const user = ledger.userById(id);
        expect(user).toEqual({id, name: "foo", aliases: [a1]});
        expect(ledger.actionLog()).toEqual([
          {
            type: "CREATE_USER",
            version: 1,
            timestamp: 0,
            userId: id,
            username: "foo",
          },
          {
            type: "ADD_ALIAS",
            version: 1,
            timestamp: 1,
            userId: id,
            alias: a1,
          },
        ]);
      });
      it("errors if there's no matching user", () => {
        const ledger = new Ledger();
        failsWithoutMutation(
          ledger,
          (l) => l.addAlias(randomUuid(), a1),
          "addAlias: no matching userId"
        );
      });
      it("throws an error if the user already has that alias", () => {
        const ledger = new Ledger();
        const id = ledger.createUser("foo");
        ledger.addAlias(id, a1);
        const thunk = () => ledger.addAlias(id, a1);
        failsWithoutMutation(ledger, thunk, "user already has alias");
      });
      it("errors if the address is another user's alias", () => {
        const ledger = new Ledger();
        const id1 = ledger.createUser("foo");
        const id2 = ledger.createUser("bar");
        ledger.addAlias(id1, a1);
        const thunk = () => ledger.addAlias(id2, a1);
        failsWithoutMutation(
          ledger,
          thunk,
          `addAlias: alias ${NodeAddress.toString(a1)} already bound`
        );
      });
      it("errors if the address is the user's innate address", () => {
        const ledger = new Ledger();
        const id = ledger.createUser("foo");
        const innateAddress = userAddress(id);
        const thunk = () => ledger.addAlias(id, innateAddress);
        failsWithoutMutation(
          ledger,
          thunk,
          `addAlias: alias ${NodeAddress.toString(innateAddress)} already bound`
        );
      });
      it("errors if the address is another user's innate address", () => {
        const ledger = new Ledger();
        const id1 = ledger.createUser("foo");
        const innateAddress = userAddress(id1);
        const id2 = ledger.createUser("bar");
        const thunk = () => ledger.addAlias(id2, innateAddress);
        failsWithoutMutation(
          ledger,
          thunk,
          `addAlias: alias ${NodeAddress.toString(innateAddress)} already bound`
        );
      });
      it("throws on unrecognized version", () => {
        // $FlowExpectedError
        expect(() => new Ledger()._addAlias({version: 1337})).toThrowError(
          "unrecognized version"
        );
      });
    });
    describe("removeAlias", () => {
      it("works", () => {
        const ledger = new Ledger();
        setFakeDate(0);
        const id = ledger.createUser("foo");
        setFakeDate(1);
        ledger.addAlias(id, a1);
        setFakeDate(2);
        ledger.removeAlias(id, a1, 0);
        const user = ledger.userById(id);
        expect(user).toEqual({id, name: "foo", aliases: []});
        expect(ledger.actionLog()).toEqual([
          {
            type: "CREATE_USER",
            version: 1,
            timestamp: 0,
            userId: id,
            username: "foo",
          },
          {
            type: "ADD_ALIAS",
            version: 1,
            timestamp: 1,
            userId: id,
            alias: a1,
          },
          {
            type: "REMOVE_ALIAS",
            version: 1,
            timestamp: 2,
            userId: id,
            alias: a1,
            retroactivePaid: "0",
          },
        ]);
      });
      it("errors if there's no matching user", () => {
        const ledger = new Ledger();
        failsWithoutMutation(
          ledger,
          (l) => l.removeAlias(randomUuid(), a1, 0),
          "removeAlias: no user matching id"
        );
      });
      it("throws an error if the user doesn't already has that alias", () => {
        const ledger = new Ledger();
        const id = ledger.createUser("foo");
        const thunk = () => ledger.removeAlias(id, a1, 0);
        failsWithoutMutation(ledger, thunk, "user does not have alias");
      });
      it("errors if the address is the user's innate address", () => {
        const ledger = new Ledger();
        const id = ledger.createUser("foo");
        const innateAddress = userAddress(id);
        const thunk = () => ledger.removeAlias(id, innateAddress, 0);
        failsWithoutMutation(
          ledger,
          thunk,
          `removeAlias: cannot remove user's innate address`
        );
      });
      it("frees the alias to be re-added", () => {
        const ledger = new Ledger();
        const id1 = ledger.createUser("foo");
        const id2 = ledger.createUser("bar");
        ledger.addAlias(id1, a1);
        ledger.removeAlias(id1, a1, 0);
        ledger.addAlias(id2, a1);
        const u2 = ledger.userById(id2);
        expect(u2).toEqual({id: id2, name: "bar", aliases: [a1]});
      });
      it("errors on invalid credProportion", () => {
        const ledger = new Ledger();
        const id1 = ledger.createUser("foo");
        ledger.addAlias(id1, a1);
        for (const bad of [-0.3, 1.3, Infinity, NaN, -Infinity]) {
          failsWithoutMutation(
            ledger,
            () => ledger.removeAlias(id1, a1, bad),
            "invalid credProportion"
          );
        }
      });
      it("throws on unrecognized version", () => {
        // $FlowExpectedError
        expect(() => new Ledger()._removeAlias({version: 1337})).toThrowError(
          "unrecognized version"
        );
      });
    });
  });

  describe("canonicalAddress", () => {
    it("users' addresses are canonical", () => {
      const ledger = new Ledger();
      const id = ledger.createUser("foo");
      const addr = userAddress(id);
      expect(ledger.canonicalAddress(addr)).toEqual(addr);
    });
    it("hitherto unseen addresses are canonical", () => {
      const ledger = new Ledger();
      expect(ledger.canonicalAddress(a1)).toEqual(a1);
    });
    it("aliases are not canonical", () => {
      const ledger = new Ledger();
      const id = ledger.createUser("foo");
      ledger.addAlias(id, a1);
      const addr = userAddress(id);
      expect(ledger.canonicalAddress(a1)).toEqual(addr);
    });
    it("unlinked aliases are again canonical", () => {
      const ledger = new Ledger();
      const id = ledger.createUser("foo");
      ledger.addAlias(id, a1).removeAlias(id, a1, 0);
      expect(ledger.canonicalAddress(a1)).toEqual(a1);
    });
  });

  describe("grain accounts", () => {
    it("newly created users have an empty account", () => {
      const ledger = new Ledger();
      const userId = ledger.createUser("foo");
      const address = userAddress(userId);
      const account = ledger.accountByAddress(address);
      expect(account).toEqual({
        userId,
        address,
        paid: "0",
        balance: "0",
      });
      expect(ledger.accounts()).toEqual([account]);
    });
    it("unseen addresses don't have accounts", () => {
      const ledger = new Ledger();
      expect(ledger.accountByAddress(a1)).toEqual(undefined);
      expect(ledger.accounts()).toEqual([]);
    });
    it("non-user addresses can have accounts", () => {
      const ledger = new Ledger();
      ledger._allocateGrain(a1, g("10"));
      const account = ledger.accountByAddress(a1);
      expect(account).toEqual({
        userId: null,
        address: a1,
        balance: "10",
        paid: "10",
      });
      expect(ledger.accounts()).toEqual([account]);
    });
    it("accountByAddress returns canonical accounts", () => {
      const ledger = new Ledger();
      const userId = ledger.createUser("foo");
      const addr = userAddress(userId);
      ledger.addAlias(userId, a1);
      expect(ledger.accountByAddress(a1)).toEqual({
        // Note: we asked for `a1`, but got address `addr`.
        // This is intended behavior.
        address: addr,
        userId,
        paid: "0",
        balance: "0",
      });
    });
    it("when a user gets an alias, it claims the alias's balance", () => {
      const ledger = new Ledger();
      const userId = ledger.createUser("foo");
      const addr = userAddress(userId);
      ledger._allocateGrain(a1, g("1"));
      ledger._allocateGrain(addr, g("1"));
      ledger.addAlias(userId, a1);
      const userAccount = ledger.accountByAddress(addr);
      expect(userAccount).toEqual({
        userId,
        address: addr,
        paid: "2",
        balance: "2",
      });
      expect(ledger.accounts()).toEqual([userAccount]);
    });
    it("if an alias is allocated Grain, it's received by the canonical address", () => {
      const ledger = new Ledger();
      const userId = ledger.createUser("foo");
      const addr = userAddress(userId);
      ledger.addAlias(userId, a1);
      ledger._allocateGrain(a1, g("1"));
      ledger._allocateGrain(addr, g("1"));
      const userAccount = ledger.accountByAddress(addr);
      expect(userAccount).toEqual({
        userId,
        address: addr,
        paid: "2",
        balance: "2",
      });
      expect(ledger.accounts()).toEqual([userAccount]);
    });
    it("removed alias accounts with retroactive payouts are handled correctly", () => {
      const examples = [
        {
          credProportion: 0.1,
          expectedRetroactivePaid: g("10"),
        },
        {
          credProportion: 1,
          expectedRetroactivePaid: g("100"),
        },
      ];
      for (const {credProportion, expectedRetroactivePaid} of examples) {
        const ledger = new Ledger();
        const userId = ledger.createUser("foo");
        const addr = userAddress(userId);
        ledger.addAlias(userId, a1);
        ledger._allocateGrain(a1, g("100"));
        ledger.removeAlias(userId, a1, credProportion);
        const aliasAccount = ledger.accountByAddress(a1);
        expect(aliasAccount).toEqual({
          userId: null,
          address: a1,
          paid: expectedRetroactivePaid,
          balance: "0",
        });
        const userAccount = ledger.accountByAddress(addr);
        expect(userAccount).toEqual({
          userId,
          address: addr,
          paid: G.sub(g("100"), expectedRetroactivePaid),
          balance: "100",
        });
        expect(ledger.accounts()).toEqual([userAccount, aliasAccount]);
      }
    });
    it("removed aliases without retroactive payouts don't have an alias account", () => {
      const ledger = new Ledger();
      const userId = ledger.createUser("foo");
      const addr = userAddress(userId);
      ledger.addAlias(userId, a1);
      ledger._allocateGrain(a1, g("100"));
      ledger.removeAlias(userId, a1, 0);
      const aliasAccount = ledger.accountByAddress(a1);
      expect(aliasAccount).toEqual(undefined);
      const userAccount = ledger.accountByAddress(addr);
      expect(userAccount).toEqual({
        userId,
        address: addr,
        paid: "100",
        balance: "100",
      });
      expect(ledger.accounts()).toEqual([userAccount]);
    });
  });

  describe("grain updates", () => {
    describe("distributeGrain", () => {
      // Motivation for the following history:
      // Across the history, both addresses earn equally, so the BALANCED
      // payout wants to give equal Grain
      // In the most recent week, only a2 had cred, so the IMMEDIATE payout
      // gives Grain only to them.
      const credHistory = deepFreeze([
        {
          cred: new Map([
            [a1, 2],
            [a2, 1],
          ]),
          intervalEndMs: 1,
        },
        {
          cred: new Map([
            [a1, 0],
            [a2, 1],
          ]),
          intervalEndMs: 3,
        },
      ]);
      it("errors on an empty credHistory", () => {
        const ledger = new Ledger();
        failsWithoutMutation(
          ledger,
          () => ledger.distributeGrain([], []),
          "distributeGrain: empty cred history"
        );
      });
      it("produces an empty distribution if there are no policies", () => {
        setFakeDate(4);
        const ledger = new Ledger().distributeGrain([], credHistory);
        expect(ledger.actionLog()).toEqual([
          {
            type: "DISTRIBUTE_GRAIN",
            version: 1,
            timestamp: 4,
            credTimestamp: 3,
            allocations: [],
          },
        ]);
        expect(ledger.accounts()).toEqual([]);
      });
      it("errors if credHistory contains non-canonical addresses", () => {
        const ledger = new Ledger();
        const id = ledger.createUser("foo");
        ledger.addAlias(id, a1);
        const thunk = () => ledger.distributeGrain([], credHistory);
        failsWithoutMutation(
          ledger,
          thunk,
          "non-canonical address in credHistory"
        );
      });
      it("computes an IMMEDIATE allocation correctly", () => {
        const policy: DistributionPolicy = {
          budget: g("10"),
          strategy: {type: "IMMEDIATE", version: 1},
        };
        setFakeDate(4);
        const ledger = new Ledger().distributeGrain([policy], credHistory);
        const account1 = ledger.accountByAddress(a1);
        const account2 = ledger.accountByAddress(a2);
        expect(account1).toEqual({
          address: a1,
          userId: null,
          balance: "0",
          paid: "0",
        });
        expect(account2).toEqual({
          address: a2,
          userId: null,
          balance: "10",
          paid: "10",
        });
        expect(ledger.actionLog()).toEqual([
          {
            type: "DISTRIBUTE_GRAIN",
            version: 1,
            credTimestamp: 3,
            timestamp: 4,
            allocations: [
              {
                strategy: policy.strategy,
                budget: policy.budget,
                version: 1,
                receipts: [
                  {address: a1, amount: "0"},
                  {address: a2, amount: "10"},
                ],
              },
            ],
          },
        ]);
      });
      it("computes a BALANCED allocation correctly", () => {
        const policy: DistributionPolicy = {
          budget: g("15"),
          strategy: {type: "BALANCED", version: 1},
        };
        setFakeDate(4);
        const ledger = new Ledger();
        // Give a2 some past payouts, so that the BALANCED
        // strategy should preferentially pay a1
        ledger._allocateGrain(a2, g("5"));
        ledger.distributeGrain([policy], credHistory);
        const account1 = ledger.accountByAddress(a1);
        const account2 = ledger.accountByAddress(a2);
        expect(account1).toEqual({
          address: a1,
          userId: null,
          balance: "10",
          paid: "10",
        });
        expect(account2).toEqual({
          address: a2,
          userId: null,
          balance: "10",
          paid: "10",
        });
        expect(ledger.actionLog()).toEqual([
          {
            type: "DISTRIBUTE_GRAIN",
            version: 1,
            credTimestamp: 3,
            timestamp: 4,
            allocations: [
              {
                strategy: policy.strategy,
                budget: policy.budget,
                version: 1,
                receipts: [
                  {address: a1, amount: "10"},
                  {address: a2, amount: "5"},
                ],
              },
            ],
          },
        ]);
      });
      it("order of the policies doesn't matter", () => {
        // We are not incrementally computing each allocation, applying it, and
        // then computing the next one.
        // Instead, we compute them all upfront, and then apply them.
        // This means that in the example below, the "BALANCED" policy doesn't
        // "know" about the payouts from the immediate policy, so someone can
        // get overpaid from the perspective of the balanced policy. This is OK
        // because it will get evened out by next week's BALANCED policy taking
        // this into account.
        const p1: DistributionPolicy = {
          budget: g("10"),
          strategy: {type: "BALANCED", version: 1},
        };
        const p2: DistributionPolicy = {
          budget: g("10"),
          strategy: {type: "IMMEDIATE", version: 1},
        };
        const ps1 = [p1, p2];
        const ps2 = [p2, p1];
        const l1 = new Ledger().distributeGrain(ps1, credHistory);
        const l2 = new Ledger().distributeGrain(ps2, credHistory);
        expect(l1.accounts()).toEqual(l2.accounts());
      });
      it("errors on unknown action version", () => {
        expect(() =>
          // $FlowExpectedError
          new Ledger()._distributeGrain({version: 1337})
        ).toThrowError("unknown DISTRIBUTE_GRAIN version: 1337");
      });
      it("errors on unknown allocation version", () => {
        expect(() =>
          new Ledger()._distributeGrain({
            version: 1,
            type: "DISTRIBUTE_GRAIN",
            timestamp: 4,
            credTimestamp: 9,
            // $FlowExpectedError
            allocations: [{version: 1337}],
          })
        ).toThrowError("unknown allocation version 1337");
      });
      it("BALANCED strategy accounts for unlinked aliases' retroactive paid", () => {
        // Sanity check since this property is important.
        const p: DistributionPolicy = {
          budget: g("7"),
          strategy: {type: "BALANCED", version: 1},
        };
        const ledger = new Ledger();
        setFakeDate(4);
        const userId = ledger.createUser("user");
        const aU = userAddress(userId);
        setFakeDate(5);
        ledger.addAlias(userId, a1);
        ledger._allocateGrain(aU, g("10"));
        setFakeDate(6);
        ledger.removeAlias(userId, a1, 0.5);
        setFakeDate(7);
        ledger.distributeGrain([p], credHistory);
        // From the perspective of the balanced payout: a1 and a2 have equal cred,
        // and a1 has a retroactive allotment of 5 Grain, so therefore from this payout,
        // 6 should go to a2, and 1 should go to a1, bringing them both to 6 Grain post-payout.
        const accountU = ledger.accountByAddress(aU);
        const account1 = ledger.accountByAddress(a1);
        const account2 = ledger.accountByAddress(a2);
        expect(accountU).toEqual({
          address: aU,
          userId,
          balance: "10",
          paid: "5",
        });
        expect(account1).toEqual({
          address: a1,
          userId: null,
          balance: "1",
          paid: "6",
        });
        expect(account2).toEqual({
          address: a2,
          userId: null,
          balance: "6",
          paid: "6",
        });
        expect(ledger.actionLog()).toEqual([
          {
            type: "CREATE_USER",
            username: "user",
            version: 1,
            timestamp: 4,
            userId,
          },
          {
            type: "ADD_ALIAS",
            version: 1,
            timestamp: 5,
            userId,
            alias: a1,
          },
          {
            type: "REMOVE_ALIAS",
            version: 1,
            timestamp: 6,
            userId,
            alias: a1,
            retroactivePaid: "5",
          },
          {
            type: "DISTRIBUTE_GRAIN",
            version: 1,
            credTimestamp: 3,
            timestamp: 7,
            allocations: [
              {
                strategy: p.strategy,
                budget: p.budget,
                version: 1,
                receipts: [
                  {address: a1, amount: "1"},
                  {address: a2, amount: "6"},
                ],
              },
            ],
          },
        ]);
        expect(ledger.accounts()).toEqual([accountU, account1, account2]);
      });
    });
  });

  describe("state reconstruction", () => {
    it("fromActionLog with an empty action log results in an empty ledger", () => {
      const emptyLog = new Ledger().actionLog();
      expect(emptyLog).toEqual([]);
      expect(Ledger.fromActionLog(emptyLog)).toEqual(new Ledger());
    });
    it("actionLog and fromActionLog compose to identity", () => {
      const ledger = new Ledger();
      const id1 = ledger.createUser("foo");
      const id2 = ledger.createUser("bar");
      ledger.addAlias(id1, a1);
      ledger.removeAlias(id1, a1, 0);
      ledger.addAlias(id2, a1);

      const ua1 = userAddress(id1);
      const ua2 = userAddress(id2);
      const policies = [
        {budget: g("400"), strategy: {type: "BALANCED", version: 1}},
        {budget: g("100"), strategy: {type: "IMMEDIATE", version: 1}},
      ];
      const credHistory = [
        {
          intervalEndMs: 10,
          cred: new Map([
            [ua1, 5],
            [ua2, 0],
          ]),
        },
        {
          intervalEndMs: 20,
          cred: new Map([
            [ua1, 3],
            [ua2, 7],
          ]),
        },
      ];
      ledger.distributeGrain(policies, credHistory);
      expect(Ledger.fromActionLog(ledger.actionLog())).toEqual(ledger);
    });
  });
});
