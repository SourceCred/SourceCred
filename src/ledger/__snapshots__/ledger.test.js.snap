// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ledger/ledger state reconstruction serialized ledger snapshots as expected 1`] = `
"[
{\\"action\\":{\\"identity\\":{\\"address\\":\\"N\\\\u0000sourcecred\\\\u0000core\\\\u0000IDENTITY\\\\u0000USER\\\\u0000YVZhbGlkVXVpZEF0TGFzdA\\\\u0000\\",\\"aliases\\":[],\\"id\\":\\"YVZhbGlkVXVpZEF0TGFzdA\\",\\"name\\":\\"steven\\",\\"subtype\\":\\"USER\\"},\\"type\\":\\"CREATE_IDENTITY\\",\\"version\\":\\"1\\"},\\"ledgerTimestamp\\":1,\\"version\\":\\"1\\"},
{\\"action\\":{\\"identity\\":{\\"address\\":\\"N\\\\u0000sourcecred\\\\u0000core\\\\u0000IDENTITY\\\\u0000ORGANIZATION\\\\u0000URgLrCxgvjHxtGJ9PgmckQ\\\\u0000\\",\\"aliases\\":[],\\"id\\":\\"URgLrCxgvjHxtGJ9PgmckQ\\",\\"name\\":\\"crystal-gems\\",\\"subtype\\":\\"ORGANIZATION\\"},\\"type\\":\\"CREATE_IDENTITY\\",\\"version\\":\\"1\\"},\\"ledgerTimestamp\\":2,\\"version\\":\\"1\\"},
{\\"action\\":{\\"alias\\":\\"N\\\\u0000a1\\\\u0000\\",\\"identityId\\":\\"YVZhbGlkVXVpZEF0TGFzdA\\",\\"type\\":\\"ADD_ALIAS\\",\\"version\\":\\"1\\"},\\"ledgerTimestamp\\":3,\\"version\\":\\"1\\"},
{\\"action\\":{\\"distribution\\":{\\"allocations\\":[{\\"id\\":\\"yYNur0NEEkh7fMaUn6n9QQ\\",\\"policy\\":{\\"budget\\":\\"100\\",\\"policyType\\":\\"IMMEDIATE\\"},\\"receipts\\":[{\\"amount\\":\\"50\\",\\"id\\":\\"YVZhbGlkVXVpZEF0TGFzdA\\"},{\\"amount\\":\\"50\\",\\"id\\":\\"URgLrCxgvjHxtGJ9PgmckQ\\"}]}],\\"credTimestamp\\":1,\\"id\\":\\"f9xPz9YGH0PuBpPAg2824Q\\"},\\"type\\":\\"DISTRIBUTE_GRAIN\\",\\"version\\":\\"1\\"},\\"ledgerTimestamp\\":4,\\"version\\":\\"1\\"},
{\\"action\\":{\\"amount\\":\\"10\\",\\"from\\":\\"YVZhbGlkVXVpZEF0TGFzdA\\",\\"memo\\":null,\\"to\\":\\"URgLrCxgvjHxtGJ9PgmckQ\\",\\"type\\":\\"TRANSFER_GRAIN\\",\\"version\\":\\"1\\"},\\"ledgerTimestamp\\":5,\\"version\\":\\"1\\"}
]"
`;
